File: Android/src/main/java/com/example/Picotador/MainActivity.java
package com.example.Picotador;


import android.app.Activity;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.VideoView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.arthenica.mobileffmpeg.FFmpeg;
import com.arthenica.mobileffmpeg.FFmpegExecution;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.List;
import java.util.Objects;

import static android.os.Environment.DIRECTORY_MOVIES;
import static android.os.Environment.getExternalStoragePublicDirectory;
import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static com.example.Picotador.Constant.allMediaList;
import static com.example.Picotador.ListFolder.load_Directory_Files;
import static com.example.Picotador.RecyclerFolderAdapter.allMedia;
import static com.example.Picotador.RecyclerFolderAdapter.count;
import static com.example.Picotador.RecyclerFolderAdapter.holderPositionClicked;
import static com.example.Picotador.RecyclerVideoAdapter.videoCount;
import static com.example.Picotador.StoriesCut.Cut;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private RecyclerFolderAdapter recyclerViewAdapter;
    private RecyclerVideoAdapter recyclerVideoAdapter;
    private Uri fileUri;
    public MenuItem delete, share, backButton, itemSelected, config, sobre;


    public MainActivity() {
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);




        //Setting Permissions
        String[] permission = {"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"};
        requestPermissions(permission, 1);


    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.items, menu);

        delete = menu.findItem(R.id.delete);
        share = menu.findItem(R.id.share);
        backButton = menu.findItem(R.id.backButton);
        itemSelected = menu.findItem(R.id.itemSelected);
        config = menu.findItem(R.id.config);
        sobre = menu.findItem(R.id.sobre);
        config.setVisible(false);
        sobre.setVisible(false);

        //RecyclerView
        recyclerView = findViewById(R.id.recyclerviewVideo);
        recyclerView.setLayoutManager(new GridLayoutManager(getApplicationContext(), 2));


        recyclerVideoAdapter = new RecyclerVideoAdapter(this, delete, share, backButton, recyclerView, itemSelected);
        recyclerVideoAdapter.setHasStableIds(true);
        recyclerViewAdapter = new RecyclerFolderAdapter(this, delete, share, backButton, recyclerView, itemSelected);
        recyclerViewAdapter.setHasStableIds(true);

        recyclerView.setAdapter(recyclerViewAdapter);


        backButton.setOnMenuItemClickListener(new MenuItem.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                if (allMediaList.get(0).toString().contains(".mp4")) {
                    if (!delete.isVisible()) {
                        allMediaList.clear();
                        allMedia.clear();
                        load_Directory_Files(new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories"));

                        RecyclerFolderAdapter recyclerFolderAdapter = new RecyclerFolderAdapter(getApplicationContext(), menu.findItem(R.id.delete), menu.findItem(R.id.share), menu.findItem(R.id.backButton), recyclerView, menu.findItem(R.id.itemSelected));
                        recyclerView.setAdapter(recyclerFolderAdapter);
                        recyclerView.setLayoutManager(new GridLayoutManager(getApplicationContext(), 2));

                        recyclerFolderAdapter.notifyDataSetChanged();

                        menu.findItem(R.id.backButton).setVisible(false);


                    } else {
                        menu.findItem(R.id.delete).setVisible(false);
                        menu.findItem(R.id.share).setVisible(false);
                        menu.findItem(R.id.itemSelected).setVisible(false);
                        findViewById(R.id.thumb).setLongClickable(true);
                        findViewById(R.id.thumb).setSelected(false);
                        count = 0;
                        videoCount = 0;
                        holderPositionClicked.clear();
                    }


                } else {
                    menu.findItem(R.id.delete).setVisible(false);
                    menu.findItem(R.id.share).setVisible(false);
                    menu.findItem(R.id.backButton).setVisible(false);
                    menu.findItem(R.id.itemSelected).setVisible(false);
                    count = 0;
                    videoCount = 0;
                    holderPositionClicked.clear();
                    findViewById(R.id.thumb).setLongClickable(true);
                    findViewById(R.id.thumb).setSelected(false);
                }

                return true;
            }
        });


        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 1) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                File dataFolder = new File(String.valueOf(getApplicationContext().getDir("cortarStories", MODE_PRIVATE)));
                File[] listFiles = dataFolder.listFiles();

                if (listFiles != null) {
                    for (File listFile : listFiles) {
                        if (listFile.isDirectory()) {
                            File[] listFolder = listFile.listFiles();
                            if (listFolder != null) {
                                for (File file : listFolder) {
                                    file.delete();
                                }
                            } else {
                                listFile.delete();
                            }

                        }
                    }
                    dataFolder.delete();
                }


                File moviesDirectory = new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories");
                if (!moviesDirectory.isDirectory()) {
                    moviesDirectory.mkdirs();
                    findViewById(R.id.selectVideo).setVisibility(VISIBLE);
                } else {
                    if (Objects.requireNonNull(moviesDirectory.listFiles()).length == 0) {
                        findViewById(R.id.selectVideo).setVisibility(VISIBLE);
                    }
                }
                File[] listMoviesFolders = new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories/").listFiles();
                if (listMoviesFolders != null) {
                    for (File listMoviesFolder : listMoviesFolders) {
                        File[] listVideo = listMoviesFolder.listFiles();
                        assert listVideo != null;
                        if (listVideo.length == 0) {
                            listMoviesFolder.delete();
                        }

                    }
                }

                File[] allFolders = new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories").listFiles();
                if (allFolders != null && allFolders.length > 0) {
                    File directory = new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories");
                    load_Directory_Files(directory);
                    if (recyclerViewAdapter.getItemCount() > 0) {
                        recyclerViewAdapter.notifyDataSetChanged();
                    }

                }


            }

        }
    }

    @Override
    public void onBackPressed() {
        if (count == 0 && allMediaList.get(0).toString().contains(".mp4")) {
            if (!delete.isVisible()) {
                allMediaList.clear();
                File directory = new File(getExternalStoragePublicDirectory(DIRECTORY_MOVIES) + "/CortarStories");
                load_Directory_Files(directory);
                recyclerView.setAdapter(recyclerViewAdapter);
                recyclerView.setLayoutManager(new GridLayoutManager(getApplicationContext(), 2));

                recyclerViewAdapter.notifyDataSetChanged();
                backButton.setVisible(false);

            } else {
                for (int i = 0; i < videoCount; i++) {
                    CheckBox checkbox = Objects.requireNonNull(Objects.requireNonNull(recyclerView.getLayoutManager()).findViewByPosition(Integer.parseInt(holderPositionClicked.get(i)))).findViewById(R.id.checkBox);
                    Objects.requireNonNull(recyclerView.getLayoutManager().findViewByPosition(Integer.parseInt(holderPositionClicked.get(i)))).findViewById(R.id.thumb).setLongClickable(true);
                    if (checkbox.isChecked()) {
                        checkbox.setChecked(false);
                        checkbox.setVisibility(INVISIBLE);
                    }
                }
                delete.setVisible(false);
                share.setVisible(false);
                backButton.setVisible(false);
                itemSelected.setVisible(false);
                findViewById(R.id.thumb).setLongClickable(true);
                holderPositionClicked.clear();
                videoCount = 0;
            }

        } else if (videoCount == 0 && !allMediaList.get(0).toString().contains(".mp4")) {
            for (int i = 0; i < count; i++) {
                CheckBox checkbox = Objects.requireNonNull(Objects.requireNonNull(recyclerView.getLayoutManager()).findViewByPosition(Integer.parseInt(holderPositionClicked.get(i)))).findViewById(R.id.checkBox);
                if (checkbox.isChecked()) {
                    checkbox.setChecked(false);
                    checkbox.setVisibility(INVISIBLE);
                }
            }


            delete.setVisible(false);
            share.setVisible(false);
            backButton.setVisible(false);
            itemSelected.setVisible(false);
            holderPositionClicked.clear();
            count = 0;
            findViewById(R.id.thumb).setLongClickable(true);
        }

    }

    public void onClick(View view) {

        if (view.equals(findViewById(R.id.select))) {
            String[] permission = {"android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE"};
            if (checkSelfPermission("READ_EXTERNAL_STORAGE") == PackageManager.PERMISSION_DENIED) {
                requestPermissions(permission, 2);
            }
            Intent intent = new Intent()
                    .setType("video/*")
                    .setAction(Intent.ACTION_GET_CONTENT);

            startActivityForResult(Intent.createChooser(intent, "Video"), 3);


        }

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1) {
            if (resultCode == Activity.RESULT_OK) {
                fileUri = data.getData();
                getContentResolver().takePersistableUriPermission(fileUri, Intent.FLAG_GRANT_WRITE_URI_PERMISSION
                        | Intent.FLAG_GRANT_READ_URI_PERMISSION);
            }
        } else if (requestCode == 5) {
            showUserSettings();

        } else if (requestCode == 3 && resultCode == Activity.RESULT_OK && data != null) {
            Uri input = data.getData();
            /*setContentView(R.layout.videoview);
            findViewById(R.id.videoView).setVisibility(VISIBLE);
            ((VideoView)findViewById(R.id.videoView)).setVideoURI(input);
            ((VideoView)findViewById(R.id.videoView)).canPause();
            ((VideoView)findViewById(R.id.videoView)).start();*/

            DialogCreate Aguarde = new DialogCreate("Aguarde", "Seu video está sendo picotado para caber no Stories.", android.R.drawable.ic_popup_sync, null, null);
            Thread StartCut = new Thread(new Runnable() {
                @Override
                public void run() {
                    StartCut(input, Aguarde);
                }

            });
            StartCut.start();
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Aguarde.show(getSupportFragmentManager(), "Aguarde");
                    findViewById(R.id.selectVideo).setVisibility(INVISIBLE);
                    findViewById(R.id.progressBar).setVisibility(VISIBLE);
                    ((ProgressBar) findViewById(R.id.progressBar)).setProgress(0);
                    findViewById(R.id.progressText).setVisibility(VISIBLE);
                    findViewById(R.id.cancelButton).setVisibility(VISIBLE);
                    findViewById(R.id.select).setClickable(false);
                }
            });

            findViewById(R.id.cancelButton).setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    StartCut.interrupt();
                    SystemClock.sleep(500);
                    List<FFmpegExecution> executions = FFmpeg.listExecutions();
                    if (executions != null && executions.size() > 0) {FFmpeg.cancel();}
                    findViewById(R.id.progressBar).setVisibility(INVISIBLE);
                    findViewById(R.id.cancelButton).setVisibility(INVISIBLE);
                    findViewById(R.id.cancelButton).setVisibility(INVISIBLE);
                    findViewById(R.id.progressText).setVisibility(INVISIBLE);
                    findViewById(R.id.select).setClickable(true);
                    File dataFolder = new File(String.valueOf(getApplicationContext().getDir("cortarStories", MODE_PRIVATE)));
                    File[] listFiles = dataFolder.listFiles();

                    if (listFiles != null) {
                        for (File listFile : listFiles) {
                            if (listFile.isDirectory()) {
                                File[] listFolder = listFile.listFiles();
                                if (listFolder != null) {
                                    for (File file : listFolder) {
                                        file.delete();
                                    }
                                } else {
                                    listFile.delete();
                                }

                            }
                        }
                        dataFolder.delete();
                    }

                    SystemClock.sleep(100);
                    getSystemService(NotificationManager.class).cancel(104);



                }
            });

        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {

            case R.id.config:
                Intent config = new Intent(this, MainActivity.class);
                startActivityForResult(config, 5);
                break;

        }

        return true;


    }
    private void showUserSettings() {
        getSupportFragmentManager().beginTransaction()
                .replace(android.R.id.content, new PicotadorSetttings())
                .commit();

    }

    private void StartCut(Uri input, DialogCreate aguarde) {


        Context context = getApplicationContext();
        String folder = (getApplicationContext().getDir("cortarStories", MODE_PRIVATE)).toString();
        Uri videoFolder = Uri.fromFile(new File(folder + "/stories.mp4"));
        //////////////////////////////////////////// Copying Selected Video //////////////////////////////////////////////////////
        InputStream in = null;
        try {
            in = getApplicationContext().getContentResolver().openInputStream(input);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        OutputStream out = null;
        try {
            out = getApplicationContext().getContentResolver().openOutputStream(videoFolder);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }

        if (in != null && out != null) {
            try {
                byte[] buf = new byte[1024];
                int len;
                while ((len = in.read(buf)) > 0) {
                    out.write(buf, 0, len);
                }

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        try {
            assert in != null;
            in.close();
            assert out != null;
            out.close();
        } catch (IOException e) {
            e.printStackTrace();
        }


        NotificationChannel channel = new NotificationChannel("Picotador", "Picotador", NotificationManager.IMPORTANCE_DEFAULT);
        getSystemService(NotificationManager.class).createNotificationChannel(channel);

        Cut(folder, context, getSupportFragmentManager(),
                recyclerViewAdapter, getSystemService(NotificationManager.class), recyclerView, aguarde, MainActivity.this);

    }

}

File: Android/src/main/java/com/example/cortarstories/MainActivity.kt
package com.example.cortarstories

import android.app.*
import android.content.ContentResolver
import android.content.Context
import android.content.DialogInterface
import android.content.DialogInterface.*
import android.content.Intent
import android.content.Intent.ACTION_MEDIA_SCANNER_SCAN_FILE
import android.content.pm.PackageManager
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.Menu
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.app.NotificationCompat
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import java.io.*
import java.lang.System.out
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.PathMatcher
import java.nio.file.Paths
import java.nio.file.Paths.get


class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        val permission = arrayOf("android.permission.READ_EXTERNAL_STORAGE", "android.permission.WRITE_EXTERNAL_STORAGE")
        requestPermissions(permission, 1)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(findViewById(R.id.toolbar))
        findViewById<FloatingActionButton>(R.id.select).setOnClickListener {
            if (checkSelfPermission("READ_EXTERNAL_STORAGE") == PackageManager.PERMISSION_DENIED) {requestPermissions(permission, 2)}
            val intent = Intent()
                    .setType("video/*")
                    .setAction(Intent.ACTION_GET_CONTENT)

            startActivityForResult(Intent.createChooser(intent, "Selecione um Video"), 111)

        }


    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        //val appDir = getDir("InstagramCut", Context.MODE_PRIVATE)

        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.dialogTitle)
        builder.setMessage(R.string.dialogMessage)
        builder.setIcon(android.R.drawable.ic_popup_sync)
        builder.setPositiveButton("OK"){ dialogInterface, which -> Toast.makeText(applicationContext, "Processando", Toast.LENGTH_SHORT).show()}
        val alertDialog: AlertDialog = builder.create()
        alertDialog.setCancelable(false)

        val builder2 = AlertDialog.Builder(this)
        builder2.setTitle(R.string.dialogTitle2)
        builder2.setMessage(R.string.dialogMessage2)
        builder2.setIcon(android.R.drawable.ic_dialog_info)
        builder2.setPositiveButton("OK"){ dialogInterface, which -> Toast.makeText(applicationContext, "Pronto", Toast.LENGTH_LONG).show()}
        val alertDialog2: AlertDialog = builder2.create()
        alertDialog2.setCancelable(false)

        val builder3 = AlertDialog.Builder(this)
        builder3.setTitle(R.string.dialogTitle3)
        builder3.setMessage(R.string.dialogMessage3)
        builder3.setIcon(android.R.drawable.ic_dialog_info)
        builder3.setPositiveButton("OK"){ dialogInterface, which -> Toast.makeText(applicationContext, "Tente Novamente", Toast.LENGTH_LONG).show()}
        val alertDialog3: AlertDialog = builder3.create()
        alertDialog3.setCancelable(false)

        if (requestCode == 111 && resultCode == RESULT_OK) {
            val input = data?.data!!
            var source: InputStream = applicationContext.contentResolver.openInputStream(input)!!
            val storiesDir = getDir("CortarStories", MODE_PRIVATE)
            var outPath = Uri.fromFile(File("$storiesDir/stories.mp4"))
            var destination: OutputStream = applicationContext.contentResolver.openOutputStream(outPath)!!
            alertDialog.show()

            val progressBar = findViewById<ProgressBar>(R.id.progressBar)
            val progressBar2 = findViewById<ProgressBar>(R.id.progressBar2)
            progressBar.visibility = View.VISIBLE
            progressBar2.visibility = View.VISIBLE
            var Snack: View = findViewById(android.R.id.content)

            var notification = NotificationCompat.Builder(this, "CortarStories")
                .setSmallIcon(R.mipmap.cortar_stories_round)
                .setContentText("Processando...")
                .setProgress(100, 0, true)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)

            var notification2 = NotificationCompat.Builder(this, "CortarStories")
                .setSmallIcon(R.mipmap.cortar_stories_round)
                .setContentText("Pronto")
                .setProgress(100, 100, false)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)



            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                val name = getString(R.string.channel_name)
                val descriptionText = getString(R.string.channel_description)
                val importance = NotificationManager.IMPORTANCE_DEFAULT
                val channel = NotificationChannel("CortarStories", name, importance).apply {
                    description = descriptionText
                }
                val notificationManager: NotificationManager =
                        getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                notificationManager.createNotificationChannel(channel)
            notificationManager.notify(103, notification.build())
            }


            val thread = Thread(Runnable {
                source.copyTo(destination)
                source.close()
                destination.close()
                var path = "$storiesDir/stories.mp4"
                var code = Stories(path, progressBar, Snack, progressBar2)
                runOnUiThread {
                    if (code == 1){
                    alertDialog.hide()
                    alertDialog2.show()
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            val name = getString(R.string.channel_name)
                            val descriptionText = getString(R.string.channel_description)
                            val importance = NotificationManager.IMPORTANCE_DEFAULT
                            val channel = NotificationChannel("CortarStories", name, importance).apply {
                                description = descriptionText
                            }
                            val notificationManager: NotificationManager =
                                getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                            notificationManager.createNotificationChannel(channel)
                            notificationManager.cancel(103)
                            notificationManager.notify(104, notification2.build())
                        }
                    }else if (code == 0){alertDialog.hide()
                        alertDialog3.show()
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                            val name = getString(R.string.channel_name)
                            val descriptionText = getString(R.string.channel_description)
                            val importance = NotificationManager.IMPORTANCE_DEFAULT
                            val channel = NotificationChannel("CortarStories", name, importance).apply {
                                description = descriptionText
                            }
                            val notificationManager: NotificationManager =
                                getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
                            notificationManager.createNotificationChannel(channel)
                            notificationManager.cancel(103)}
                    }
                }
            })

            thread.start()

        }


    }
}


